### PR TITLE
Devcontainer workspaces openable through recent workspaces

### PR DESCRIPTION
**Pull request by jonah-iden**
#### What it does

This adds devcontainer workspaces to the recent workspaces so they can be opened from there. 
Also adds the ability to add custom workspace types through contributions to theia in general

Maybe the `workspaceServer` should be a `localProxy` so that it still shows local workspaces even when connected to a remote? Any comments regarding this would be appreciated.

Implements #14292

#### How to test

Open a dev-container. After disconnecting again the workspace should appear in recent workspaces as `workspace name  devcontainer name [Dev Container]`. Selecting it should directly open the devcontainer.

#### Follow-ups

Since devcontainers is using the WorkspaceStorageService to store the last container info, its different based on what workspace is open. So if no workspace is open it will create a new container instead of reusing the initially created one.
We need some other place to store this info workspace independent.

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by sgraband** [comment]
> Maybe the workspaceServer should be a localProxy so that it still shows local workspaces even when connected to a remote? Any comments regarding this would be appreciated.

I believe that would be a great addition. This helps a lot with not having to change your workflows when working in a devconainer imho. I am however not sure what the effort is for this. But if it is a bigger effort we should definately track this in an issue :+1:

**Comment by jonah-iden** [comment]
nah thats one code line change. I'm just not sure if this could create any other issues. But i don't think so. 
i guess lets just try it out

**Comment by jonah-iden** [comment]
ok iguess its a little more work, since opening local workspaces would require disconnecting the remote

**Comment by jonah-iden** [comment]
i think it's better to do the local-workspace server in another task. Its quite a bit more work than i first thought. Also on a conceptual level since in theory you can have multiple workspaceses inside of the container and outside on the local system.

File: packages/dev-container/src/electron-browser/container-connection-contribution.ts [hunk 1]
```diff
@@ -59,46 +64,85 @@ export class ContainerConnectionContribution extends AbstractRemoteRegistryContr
...
+    }
+
+    async getWorkspaceLabel(uri: URI): Promise<string | undefined> {
+        const containerFilePath = new URLSearchParams(uri.query).get(DEV_CONTAINER_PATH_QUERY);
+        if (!containerFilePath) {
+            return;
+        };
+        const files = await this.connectionProvider.getDevContainerFiles(uri.path.toString());
+        const devcontainerFile = files.find(file => file.path === containerFilePath);
+        return `${uri.path.base} @ ${devcontainerFile?.name} [Dev Container]`;
```
Comment (sgraband) [inline packages/dev-container/src/electron-browser/container-connection-contribution.ts:95]: I personally find this way of representing the information a bit to clutered. I would prefer something like in VSCode (e.g. `{uri.path.base} [Dev Container: ${devcontainerFile?.name}]`.
But this is of course only my personal feeling. WDYT?

Also would it be possible to render the output of the `getWorkspaceLabel()` method in the title bar of the application? Currently this does not display the indicator that it is a dev container in the title for me.

**Review by sgraband — COMMENTED** [review]
Thanks for the change @jonah-iden! Great improvement and works great! I just have one inline comment and a general remark:

Currently, when the last workspace is a dev container and i open Theia again it will not bring up the dev container, but an empty workspace. Is there something we can do about that?

**Comment by sgraband** [comment]
> i think it's better to do the local-workspace server in another task. Its quite a bit more work than i first thought. Also on a conceptual level since in theory you can have multiple workspaceses inside of the container and outside on the local system.


Yes i agree, could you create a ticket for this with your finding or should i take this over?

**Comment by jonah-iden** [comment]
sure, i'll create a ticket

File: packages/dev-container/src/electron-browser/container-connection-contribution.ts (hunk 1 above)
Comment (jonah-iden) [reply packages/dev-container/src/electron-browser/container-connection-contribution.ts:95]: This is exactly how its displayed for me in vscode. So i just adapted that. But im totally open to any other formats

Regarding your second point thats something Jan is doing in this [PR](https://github.com/eclipse-theia/theia/pull/14571#event-15506041995)

**Comment by jonah-iden** [comment]
created an issue [here](https://github.com/eclipse-theia/theia/issues/14574)

**Review by sgraband — APPROVED** [review]
Thanks @jonah-iden! I think this label is much more readable :tada:

**Review by sgraband — COMMENTED** [review]
> Currently, when the last workspace is a dev container and i open Theia again it will not bring up the dev container, but an empty workspace. Is there something we can do about that?

I forgot about this. Do you think this is easily fixable or should we create a follow up for this as well?

**Comment by jonah-iden** [comment]
its kind of part of the other issue i created recently which would allow opening local workspaces from when connected to a remote or container. At least its not super simple to do but im already looking into the new issue so i can take a look here as well

**Pull request merged**